pkg/clients/gcp/network/network.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package network
+
+import (
+	googlecompute "google.golang.org/api/compute/v1"
+
+	"github.com/crossplaneio/crossplane/gcp/apis/compute/v1alpha1"
+)
+
+// GenerateNetwork takes a *GCPNetworkSpec and returns *googlecompute.Network.
```

Comment:
It is my opinion that all "translation" between Crossplane and a provider API should be in the `client` pkg. Following that logic, I would prefer functions such as [`IsSameAs()`](https://github.com/crossplaneio/crossplane/pull/699/files#diff-3febc70e3e5058903993e97b62dc51cfR157) to also exist in `client`. However, I do see that there is some synergies with having it be an actual method on the `Spec`, so maybe I am wrong in my thinking. In general though, I like to think of the `client` pkg as a "translation layer" where all conversion and comparison takes place. Otherwise, we don't really need the `client` pkg if we are calling the provider API directly in the controllers and doing all translation in the pkg where the Crossplane API definitions live.

Reply:
I don't know exactly where, but I think we discussed that we try to remove helper functions and ideally get rid of the `client` folder, which makes sense to me. I believe, in the end, if we have `apis` folder that represents our public interface and `controller` folder(s) where we do our operations(including translation), the project will be easier to digest and browse in general. The more contained each resource is the easier development we have, hence lower barrier for new community members.

The actual technical reason for removing translation code of google<->crossplane that @negz brought up, which I find very compelling, is that our api structs are going to be imported by other packages and it's better that we don't force those developers to import provider's SDK just because our translation code lives there. An outstanding use case for that would be that when a developer implements a stack controller for their application, they might need to import our api structs for interaction with CRs and having the provider's SDK as dependency for them just doesn't make sense.

That ended up being a long comment, but what I'm saying is that I think we should strive for contained resource code and drift when we have a good reason to (such as dependency thing mentioned above). It might be time to raise an issue about folder structure as we move forward with stack repos. The earlier to get that right in the stack repos, the better.

Reply:
Hmm I am not familiar with an effort towards removing the `client` pkg. Are you thinking of our desire to remove `util`? Regardless, I think you make some very valid points, and I would be willing to go with this structure across the board if we reach universal agreement. If `client` was completely eliminated, I assume we would just have our implementation of `newClientFn` for each managed controller in the the same file / pkg as the controller (or potentially a "parent" directory if we go with a different pkg for every managed kind).